pkg/reaper/nodereaper/nodereaper.go

```diff
@@ -551,31 +566,75 @@ func (ctx *ReaperContext) reapOldNodes(w ReaperAwsAuth) error {
 		if ctx.DryRun {
 			log.Warnf("dry run is on, '%v' will not be cordon/drained", instance.NodeName)
 		}
-		err = ctx.drainNode(instance.NodeName, ctx.DryRun)
-		if err != nil {
-			return err
-		}
 
-		err = dumpSpec(instance.NodeName, ctx.KubernetesClient)
-		if err != nil {
-			log.Warnf("failed to dump spec for node %v, %v", instance.NodeName, err)
+		var lock LockRecord
+
+		// Dry run does not need to bother with locks
+		// and only master nodes need one
+		if !ctx.DryRun && isControlPlaneNode && ctx.shouldLock() {
+			lock, err = ctx.obtainReapLock(w.DDB, instance.NodeName, instance.InstanceID, controlPlaneType)
+			if err != nil {
+				// we try to clear the lock is possible, but on failure we skip this node and continue
+				// because the lock affects only master nodes
+				// TODO: alert on long-lived locks and/or failure to clean up
+				ctx.tryClearLock(w.DDB, err, &lock)
+				continue
```

Comment:
Hmmmm, I think I might be missing something, but it seems like we are trying to clear the lock after failing to acquire it? Does a failure to acquire the lock not imply that the lock is acquired somewhere else?

That is, if I understand this correctly, `obtainReapLock` will never error if there is another governor process holding the lock and only if there is an actual AWS error?

Reply:
This is for a case where lines 628 and 590 happen. If we killed a master and the control plane did not become healthy in X minutes we'll bail, but won't delete the lock in case there's something wrong (maybe broken AMI). But, if the issue is resolved eventually, and another master in the same cluster is up for termination, `tryClearLock` will end up deleting the old lock _from this cluster_, thus unblocking the master rolls.

The reason for `continue` here is because it's way simpler code this way. We're not in a rush, we'll get to it eventually.

Reply:
Oh, also `obtainReapLock` will return an error is there's another lock - because `PutItem` would error out.